File: ruby/sum-of-multiples/sum_of_multiples.rb
# Sums multiples of 3, 5, or arbitrary numbers
class SumOfMultiples
  def initialize(*multiples)
    @multiples = multiples.any? ? multiples : [3, 5]
  end

  def self.to(num)
    SumOfMultiples.new.to(num)
  end

  def to(num)
    (1...num).select { |i| @multiples.any? { |x| i % x == 0 } }.inject(0, :+)
  end
end

File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    0..(limit-1)
    |> Enum.filter(&(any_factor?(&1, factors)))
    |> Enum.sum
  end

  defp any_factor?(n, factors), do: Enum.any?(factors, &(rem(n, &1) == 0))
end
